core/base/inc/TVirtualMutex.h

```diff
@@ -110,8 +110,15 @@ class TLockGuard {
 #endif
 
 #ifdef R__USE_IMT
-#define R__LOCKGUARD_IMT(mutex)  if (ROOT::Internal::IsParBranchProcessingEnabled()) R__LOCKGUARD(mutex)
-#define R__LOCKGUARD_IMT2(mutex) if (ROOT::Internal::IsParBranchProcessingEnabled()) R__LOCKGUARD2(mutex)
```

Comment:
Hi @pcanal , I understand you are saying that we were taking these locks even if `ROOT::Internal::IsParBranchProcessingEnabled()` was false? How can it be, given the code that was there, with an if in front?

Reply:
Well two things ... Putting a if in front of a lock guard is really bad as it says:
```
if (predicated) { LockGuard(mutex); }
```
i.e. by the time the if statement is finished (in the case of true) the lock has been acquire **AND** released and thus the following code is not protected and thus the whole thing is pointless ... [This is the case for the old R__LOCKGUARD_IMT].

The old R__LOCKGUARD_IMT2 happens to work because it actually expand to:
```
if (ROOT::Internal::IsParBranchProcessingEnabled())
   if (gGlobalMutex && !gROOTMutex) {
      gGlobalMutex->Lock(); 
      if (!gROOTMutex) gROOTMutex = gGlobalMutex->Factory(kTRUE);
      gGlobalMutex->UnLock();
   }
TLockGuard R__guard467(gROOTMutex);
```
and so only the conditional initialization (which happens to not be needed at all) is behind the IsParBranchProcessingEnabled while the lock guard itself is **unconditionally** taken and has the right/intended scope.

Cheers,
Philippe.

Reply:
Oh yes I see, the condition check was indeed ill-defined and the multi-threaded case did not crash because of the last lock guard, thanks for spotting that!
Just one more comment: `IsParBranchProcessingEnabled()` might get called twice in the new implementation, once in `R__LOCKGUARD_IMT2` and again when the latter calls `R__LOCKGUARD_IMT`. Perhaps we can just replace `R__LOCKGUARD_IMT` by `R__LOCKGUARD` at the end of `R__LOCKGUARD_IMT2`.

Reply:
Hi Philippe,
why is
```c++
R__LOCKGUARD(ROOT::Internal::IsParBranchProcessingEnabled() ? mutex : nullptr)
```
safer than
```c++
if (ROOT::Internal::IsParBranchProcessingEnabled())
   R__LOCKGUARD(mutex)
else
   R__LOCKGUARD(nullptr)
```
?

Reply:
Because we are using a Lock guard.  The code snippet above is strickly equivalent to:
```
if (ROOT::Internal::IsParBranchProcessingEnabled()) {
   if (mutex) mutex->Lock();
   if (mutex) mutex->UnLock();
} else {
   // do nothing because the argument was nullptr.
}
```
which is 'safe' but, humm, ... uninteresting

Reply:
> Perhaps we can just replace R__LOCKGUARD_IMT by R__LOCKGUARD at the end of R__LOCKGUARD_IMT2.

The double call would be rare, i.e. only when the mutex is not yet created **and** I don't think it can be done that easily;  Replacing R__LOCKGUARD_IMT by R__LOCKGUARD  would make it unconditional, wouldn't it?

Reply:
Yes, let's leave it as it is then, anyway we are now using only `R__LOCKGUARD_IMT`